### PR TITLE
Rework mssql driver to support curve geometries (#1299)

### PR DESCRIPTION
**Pull request by szekerest**

### CODE REVIEW

File: gdal/ogr/ogrsf_frmts/mssqlspatial/drv_mssqlspatial.html [hunk 1]
```diff
@@ -139,6 +139,7 @@ <h2>Configuration options</h2>
     <li><b>MSSQLSPATIAL_SHOW_FID_COLUMN</b>: Force to display the FID colums as a feature attribute. Default = "NO".</li>
     <li><b>MSSQLSPATIAL_USE_GEOMETRY_COLUMNS</b>: Use/create geometry_columns metadata table in the database. Default = "YES".</li>
     <li><b>MSSQLSPATIAL_LIST_ALL_TABLES</b>: Use mssql catalog to list available layers. Default = "NO".</li>
+    <li><b>MSSQLSPATIAL_USE_GEOMETRY_VALIDATION</b>: (From GDAL 2.5.0) Let the driver to detect the geometries which would trigger run time errors at MSSQL server. The driver tries to correct these geometries before submitting that to the server. Default = "NO".</li>
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/mssqlspatial/drv_mssqlspatial.html]: the 'to' after 'Let the driver' should be remove, no ?

File: gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometrywriter.cpp [hunk 2]
```diff
@@ -203,41 +232,144 @@ void OGRMSSQLGeometryWriter::WritePoint(double x, double y)
...
+    OGRSimpleCurve* poSubGeom;
+    WriteByte(FigureAttribute(iFigure), FA_CURVE);
     WriteInt32(PointOffset(iFigure), iPoint);
-    if (chProps & SP_HASZVALUES)
+    for (int i = 0; i < poGeom->getNumCurves(); i++)
     {
-        for (i = 0; i < poGeom->getNumPoints(); i++)
-            WritePoint(poGeom->getX(i), poGeom->getY(i), poGeom->getZ(i));
+        poSubGeom = poGeom->getCurve(i)->toSimpleCurve();
+        switch (poSubGeom->getGeometryType())
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometrywriter.cpp]: minor: you coud use          switch (wkbFlatten(poSubGeom->getGeometryType())) and use only the 2D version of the wkb constants in the case 
This holds for similar switch() in other files

File: gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometryparser.cpp [hunk 3]
```diff
@@ -31,7 +31,7 @@
 
 CPL_CVSID("$Id$")
 
-/*   SqlGeometry serialization format
+/*   SqlGeometry/SqlGeography serialization format
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometryparser.cpp]: this description is also repeated in geometrywriter.cpp. Perhaps move it in a mssqlgeometryformat.txt file ?

File: gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometrywriter.cpp (hunk 2 above)
Comment (szekerest) [reply gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometrywriter.cpp]: @rouault Do you think if the code is more readable in that case? We should also enumerate the M/ZM variants however.

File: gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometryparser.cpp (hunk 3 above)
Comment (szekerest) [reply gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometryparser.cpp]: Yes, but it was already there, however. I've just added the MSSQL2012 additions.

File: gdal/ogr/ogrsf_frmts/mssqlspatial/drv_mssqlspatial.html (hunk 1 above)
Comment (szekerest) [reply gdal/ogr/ogrsf_frmts/mssqlspatial/drv_mssqlspatial.html]: probably yes, will fix that

File: gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometrywriter.cpp (hunk 2 above)
Comment (rouault) [reply gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometrywriter.cpp]: > We should also enumerate the M/ZM variants however.

no, the code would become just
```
switch (wkbFlatten(poSubGeom->getGeometryType()))
{
        case wkbLineString:
            // do stuff
            break;
        case wkbCircularString:
            // do stuff
            break;
        default:
            break;
}
```
Looks better to me

**Pull request merged**